### ISSUE TITLE
Stackdriver and Google Cloud Logging not using new Kubernetes Engine Monitoring resource types

### ISSUE THREAD

**Issue by ppdeassis** [googleapis/google-cloud-ruby#5111]
The problem: after deploying a Rails app to GKE, I can't filter Google Cloud Logging by _resource_ (Kubernetes container) to find its specific log entries (_although I can find it by using advanced query filtering with `logName` alone_). 

#### Environment details

  - OS: N/A _(docker-alpine-3.10/3.11)_
  - Ruby version: 2.6
  - Gem name and version: `stackdriver (0.16.1)`, `rails (6.0.2.1)`

#### Steps to reproduce

1. Deploy a Rails app to GKE (`1.14.10-gke.17`) - using [Kubernetes Engine Monitoring](https://cloud.google.com/monitoring/kubernetes-engine/migration) - with the `stackdriver` gem to enable Google Cloud Logging (and other services)
2. Access Google Cloud Logging > Logs Viewer _tool_ and search for the rails app log entries by filtering by _resource_ (Kubernetes container)

You'll find out that the _monitored resource_ metadata are somehow _incomplete_ and do not follow the [new _resource type_ values](https://cloud.google.com/monitoring/kubernetes-engine/migration#resource_type_changes). So, if you try to filter the logs by resource (Kubernetes container), the search will be empty.

I'll try to list the _resource_ values from log entries below to show you what I think is the problem:

Expected values (my guess, after reading the docs):
```
resource: {
  type: "k8s_container",
  labels: { 
    project_id: "your-project-id-here",
    cluster_name: "your-cluster-name-here",
    container_name: "your-container-name-here",
    namespace_id: "your-namespace-here"
  }
}
```

Actual values (found on Logs Viewer):
```
resource: {
  type: "container"
  labels: {
    pod_id: ""
    zone: ""
    project_id: "your-project-id-here"
    cluster_name: "your-cluster-name-here"
    container_name: ""
    namespace_id: "your-namespace-here"
    instance_id: ""
  }
}
```

As you can see: 
- `resource.type` should be `k8s_container` (Kubernetes Engine Monitoring), but it's `container` (Legacy Logging and Monitoring)
- `resource.labels.container_name` is _empty_

#### Code example

- `gem "stackdriver"` in Gemfile
- `config.google_cloud.logging.log_name = "custom-log-name"` in `environments/production.rb`

#### Full backtrace

N/A

---

ps: I was in doubt to open a _feature request_ or a _bug request_. Let me know if I can change anything in the description to make it more meaningful.

**Comment by quartzmo** [comment]
@ppdeassis Thank you for opening this issue. I agree, it's hard to know how to categorize this right now (we will investigate), so I'll start by labeling it as a question.

**Comment by quartzmo** [comment]
@ppdeassis I'm pretty certain that the resource structure in [`Middleware.default_monitored_resource`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-logging/lib/google/cloud/logging/middleware.rb#L244-L250) has simply fallen out of sync with the service definition, and that it needs to be updated. It may be a week or two before we have time to work on it. Do you have an interest in attempting a pull request to fix this?

**Comment by ppdeassis** [comment]
Well, I started looking into it but I think [`Google::Cloud.env`](https://github.com/googleapis/google-cloud-ruby/blob/53560492f46dd48f5f1a8d4b9748bfeb1a4e8b5f/google-cloud-env/lib/google/cloud/env.rb) needs more data/metadata in order to discover if running app is in a Kubernetes Cluster _using_ Kubernetes Engine Monitoring (or not), because it must stay as it is now to work on clusters using "Legacy Logging and Monitoring".

If you can point me in the right direction on how to get these data/metadata I can give it a try in the meantime.

**Comment by quartzmo** [comment]
I read through [Migrating to Kubernetes Engine Monitoring](https://cloud.google.com/monitoring/kubernetes-engine/migration) but didn't see anything directly helpful. If anyone can suggest how to detect if Kubernetes Engine Monitoring is being used, please let us know.

**Comment by simonz130** [comment]
@dazuma, could you please take a look?

**Comment by matteo-rossi-wise** [comment]
@dazuma I'm experiencing the same issue, still no fix after 9 months?

**Comment by sa-orlov** [comment]
Deleted accaunt google

**Comment by matteo-rossi-wise** [comment]
**TL;DR**: remove "stackdriver" gem if you're on GKE

@ppdeassis lesson learned: If you're running your app on GKE you simply need to print to $stdout (text or JSON format) and the underlying custom fluentd agent will take care of the rest (like exporting STDOUT logs to Logs Viewer in the correct namespace)

Note: This will not solve the issue described in the main thread but can save you from headache if you need logging working in your cluster

**Comment by ppdeassis** [comment]
@matteo-rossi-wise yeah! we ended up doing that too (removing `stackdriver` gem and adapting our loggers to `STDOUT`).

**Comment by romanvbabenko** [comment]
Any updates? After 2 years the issue is still there.

**Comment by dazuma** [comment]
Apologies that this fell through the cracks. The logging library is badly out of date; we have a project to reevaluate and redesign it, and we'll take this feedback into account during that process. We'll be tracking this internally at Google for now, so I'm closing this issue from GitHub. If you have additional questions or comments feel free to reopen (or open a new issue), and reference internal project number b/407801439.

That said, I also second the comment by @matteo-rossi-wise that it may be best to log straight to stdout/stderr and let the Kubernetes Engine logging agent pick up the data from there. We've already started work on a helper library (https://github.com/googleapis/ruby-core-libraries/tree/main/google-logging-utils) that can assist with constructing the JSON-formatted logs that can be parsed by the agent. The reimagined google-cloud-logging library will be based on this foundation, but you can also use it directly.
